src/facade/dragonfly_connection.cc

```diff
@@ -2254,4 +2288,173 @@ void ResetStats() {
     io_req_size_hist->Clear();
 }
 
+bool Connection::OffloadBackpressureToDiskIfNeeded(absl::FunctionRef<MessageHandle()> handle) {
+  // Offload only when dispatch_q_ crosses watermark or when backing queue already
+  // has pending items.
+  if (backing_queue_ &&
+      ((dispatch_q_.size() > backing_queue_->Watermark()) || !backing_queue_->Empty())) {
+    auto ec = backing_queue_->Init();
+    LOG_IF(ERROR, ec) << "Failed to init disk backed backpressure with error " << ec.message();
+
+    MessageHandle msg;
+    if (!ec) {
+      msg = handle();
+      PipelineMessage* pmsg = std::get<Connection::PipelineMessagePtr>(msg.handle).get();
+      if (backing_queue_->HasEnoughBackingSpace(pmsg)) {
+        backing_queue_->OffloadToBacking(pmsg);
+        if (dispatch_q_.size() == 0) {
+          disk_backpressure_available_ = true;
+          cnd_.notify_one();
+        }
+        // Recycle message
+        QueueBackpressure& qbp = GetQueueBackpressure();
+        if (stats_->pipeline_cmd_cache_bytes < qbp.pipeline_cache_limit) {
+          stats_->pipeline_cmd_cache_bytes += pmsg->StorageCapacity();
+          pipeline_req_pool_.push_back(
+              std::move(std::get<Connection::PipelineMessagePtr>(msg.handle)));
+        }
+        // item offloaded to disk without errors, unblock connection fiber
+        return true;
+      }
+      LOG(WARNING) << "Disk backpressure file size limit reached. Could not offload backpressure.";
+    }
+  }
+  return false;
+}
+
+void Connection::LoadBackpressureFromDiskIfNeeded() {
+  if (HasDiskBacked()) {
+    auto q_insert_cb = [this](io::MutableBytes bytes) {
+      PipelineMessagePtr ptr;
+      if (ptr = GetFromPipelinePool(); ptr) {
+        ptr->storage.resize(bytes.size());
+      } else {
+        ptr = make_unique<PipelineMessage>(1, 1);
+        ptr->storage.resize(bytes.size());
+      }
+
+      memcpy(ptr->storage.begin(), bytes.begin(), bytes.size());
+      std::string_view read{reinterpret_cast<char*>(ptr->storage.data()), bytes.size()};
+      ptr->args = absl::StrSplit(read, '\0', absl::SkipEmpty());
+
+      SendAsync({.handle = std::move(ptr)});
+    };
+
+    backing_queue_->LoadFromDiskToQueue(q_insert_cb);
+  }
+}
+
+size_t Connection::DiskBackedBackpressureQueue::unique_id = 0;
+
+Connection::DiskBackedBackpressureQueue::DiskBackedBackpressureQueue()
+    : max_backing_size_(absl::GetFlag(FLAGS_connection_disk_backpressure_file_max_bytes)),
+      max_queue_load_size_(absl::GetFlag(FLAGS_connection_disk_backpressure_load_size)),
+      watermark_(absl::GetFlag(FLAGS_connection_disk_backpressure_watermark)) {
+  id_ = ++unique_id;
```

Comment:
this is buggy as `DiskBackedBackpressureQueue` is initialized in multiple threads.

Reply:
yeah it's a data race indeed. I wanted a unique identifier. will fix